application/src/ext/java/org/opentripplanner/ext/carpooling/routing/InsertionEvaluator.java

```diff
@@ -0,0 +1,343 @@
+      } else {
+        // This segment involves passenger - route it
+        var fromCoord = modifiedPoints.get(i);
+        var toCoord = modifiedPoints.get(i + 1);
+        GenericLocation from = GenericLocation.fromCoordinate(
+          fromCoord.latitude(),
+          fromCoord.longitude()
+        );
+        GenericLocation to = GenericLocation.fromCoordinate(
+          toCoord.latitude(),
+          toCoord.longitude()
+        );
+
+        segment = routingFunction.route(from, to);
+        if (segment == null) {
+          LOG.trace("Routing failed for new segment {} → {}", i, i + 1);
+          return null;
+        }
+        LOG.trace("Routed new segment for modified position {}", i);
+      }
+
+      segments.add(segment);
+    }
+
+    return segments;
+  }
+
+  /**
+   * Maps a modified route segment index to the corresponding baseline segment index.
+   * Returns -1 if the segment cannot be reused (endpoints don't match).
+   *
+   * <p>A baseline segment can only be reused if BOTH endpoints match exactly between
+   * the baseline and modified routes. This ensures we don't reuse a segment whose
+   * endpoints have changed due to passenger insertion.
+   *
+   * @param modifiedIndex Index in modified route (with passenger inserted)
+   * @param originalPoints Original route points (before passenger insertion)
+   * @param modifiedPoints Modified route points (after passenger insertion)
+   * @return Baseline segment index if endpoints match, or -1 if segment must be routed
+   */
+  private int getBaselineSegmentIndex(
+    int modifiedIndex,
+    List<WgsCoordinate> originalPoints,
+    List<WgsCoordinate> modifiedPoints
+  ) {
+    // Get the start and end coordinates of this modified segment
+    WgsCoordinate modifiedStart = modifiedPoints.get(modifiedIndex);
+    WgsCoordinate modifiedEnd = modifiedPoints.get(modifiedIndex + 1);
+
+    // Search through baseline segments to find one with matching endpoints
+    for (int baselineIndex = 0; baselineIndex < originalPoints.size() - 1; baselineIndex++) {
+      WgsCoordinate baselineStart = originalPoints.get(baselineIndex);
+      WgsCoordinate baselineEnd = originalPoints.get(baselineIndex + 1);
+
+      // Check if both endpoints match (using WgsCoordinate's built-in equality)
+      if (modifiedStart.equals(baselineStart) && modifiedEnd.equals(baselineEnd)) {
+        LOG.trace(
+          "Modified segment {} matches baseline segment {} (endpoints match)",
+          modifiedIndex,
+          baselineIndex
+        );
+        return baselineIndex;
+      }
+    }
+
+    LOG.trace(
+      "Modified segment {} has no matching baseline segment (endpoints changed)",
+      modifiedIndex
+    );
+    return -1;
+  }
+
+  /**
+   * Functional interface for street routing.
+   */
+  @FunctionalInterface
+  public interface RoutingFunction {
+    GraphPath<State, Edge, Vertex> route(GenericLocation from, GenericLocation to);
+  }
```

Comment:
This probably should be in a separate class.

Reply:
We do NOT allow public inner types (enums, interfaces or classes).